core/parameters/src/config.rs

```diff
@@ -94,3 +98,111 @@ impl Default for AccountCreationConfig {
         }
     }
 }
+
+/// The configuration for congestion control.
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct CongestionControlConfig {
+    /// How much gas in delayed receipts of a shard is 100% incoming congestion.
+    ///
+    /// Based on incoming congestion levels, a shard reduces the gas it spends on
+    /// accepting new transactions instead of working on incoming receipts. Plus,
+    /// incoming congestion contributes to overall congestion, which reduces how
+    /// much other shards are allowed to forward to this shard.
+    pub max_congestion_incoming_gas: Gas,
+
+    /// How much gas in outgoing buffered receipts of a shard is 100% congested.
+    ///
+    /// Outgoing congestion contributes to overall congestion, which reduces how
+    /// much other shards are allowed to forward to this shard.
+    pub max_congestion_outgoing_gas: Gas,
+
+    /// How much memory space of all delayed and buffered receipts in a shard is
+    /// considered 100% congested.
+    ///
+    /// Memory congestion contributes to overall congestion, which reduces how much
+    /// other shards are allowed to forward to this shard.
+    ///
+    /// This threshold limits memory requirements of validators to a degree but it
+    /// is not a hard guarantee.
+    pub max_congestion_memory_consumption: u64,
+
+    /// How many missed chunks in a row in a shard is considered 100% congested.
+    /// TODO(congestion_control) - find a good limit for missed chunks.
+    pub max_congestion_missed_chunks: u64,
+
+    /// The maximum amount of gas attached to receipts a shard can forward to
+    /// another shard per chunk.
+    ///
+    /// The actual gas forwarding allowance is a linear interpolation between
+    /// [`MIN_OUTGOING_GAS`] and [`MAX_OUTGOING_GAS`], or 0 if the receiver is
+    /// fully congested.
+    pub max_outgoing_gas: Gas,
+
+    /// The minimum gas each shard can send to a shard that is not fully congested.
+    ///
+    /// The actual gas forwarding allowance is a linear interpolation between
+    /// [`MIN_OUTGOING_GAS`] and [`MAX_OUTGOING_GAS`], or 0 if the receiver is
+    /// fully congested.
+    pub min_outgoing_gas: Gas,
+
+    /// How much gas the chosen allowed shard can send to a 100% congested shard.
+    ///
+    /// This amount is the absolute minimum of new workload a congested shard has to
+    /// accept every round. It ensures deadlocks are provably impossible. But in
+    /// ideal conditions, the gradual reduction of new workload entering the system
+    /// combined with gradually limited forwarding to congested shards should
+    /// prevent shards from becoming 100% congested in the first place.
+    pub red_gas: Gas,
```

Comment:
I think we need a new name for this. "red" was from the traffic light control analogy.

Not sure about what would be a better name, though. Maybe something like `allowed_shard_gas`, `deadlock_prevention_gas`, or `min_incoming_gas`. Do you have idea?

Reply:
I renamed it to allowed_shard_outgoing_gas to keep it consistent with the other `*_outgoing_gas` limits.

Reply:
Good name!